Rust: less strict object lifetimes

mmastrac opened the pull request:
NOTE: this can't be merged until https://github.com/google/flatbuffers/pull/4925 lands.

This tweaks the rust codegen so that object lifetimes no longer require `self` to live as long as the table. That had the effect of requiring an inflated borrow of intermediate tables which prevented code like `monster.map(|e| e.enemy()).and_then(|e| e.name())` from working.

I've confirmed that this continues compiling in my (private unfortunately) Rust codebase with a pretty complex object structure, and that the base commits appear to be correct as well.

mmastrac commented:
New test passes and re-confirmed still works on my local codebase. The test I added confirms that the lifetime of the output object is disconnected from the lifetime of self, as it is required to outlive the temporary created in the `map` function.

With the old-style code, the test fails with a borrow-checker error:

```
  #[inline]
  pub fn name(&'a self) -> &'a str {
    self._tab.get::<flatbuffers::ForwardsUOffset<&str>>(Monster::VT_NAME, None).unwrap()
  }
```

```
error[E0597]: `e` does not live long enough
   --> tests/integration_test.rs:273:57
    |
273 |         let enemy_of_my_enemy = monster.enemy().map(|e| e.name());
    |                                                         ^      - `e` dropped here while still borrowed
    |                                                         |
    |                                                         borrowed value does not live long enough
274 |         assert_eq!(enemy_of_my_enemy, Some("Fred"));
275 |     }
    |     - borrowed value needs to live until here
```

rw commented on tests/rust_usage_test/tests/integration_test.rs:
```diff
@@ -263,6 +263,16 @@ mod lifetime_correctness {
     }
+
+    #[test]
+    fn object_lifetime() {
```
Please make this a more descriptive test name.

mmastrac replied:
Roger, will do.

mmastrac commented:
I cleaned up some that extra `bfbs` file that got committed somehow. Also renamed the test to `table_object_self_lifetime_is_short_lived` and added more commentary around it.

rw commented on tests/rust_usage_test/tests/integration_test.rs:
```diff
@@ -263,6 +263,16 @@ mod lifetime_correctness {
+        let buf = load_file("../monsterdata_test.mon");
+        let monster = my_game::example::get_root_as_monster(&buf);
+        // This line won't compile if "self" is required to live for the lifetime of buf above as the borrow disappears at the end of the closure.
+        let enemy_of_my_enemy = monster.enemy().map(|e| e.name());
```
This would be clearer as `monster.enemy().unwrap().name()`, WDYT?

mmastrac replied:
You are correct, but in this case I'm specifically testing a failing pattern that won't compile. It'll work if there's no closure in place as the temporary consumed from the `Option` will be allowed live to the end of the function. The closure "forces the issue".

mmastrac replied:
I could modify the test to more explicitly use `{` and `}` instead of a closure. For example, this would probably fail too:

```
         let buf = load_file("../monsterdata_test.mon");
        let monster = my_game::example::get_root_as_monster(&buf);
        // This line won't compile if "self" is required to live for the lifetime of buf above as the borrow disappears at the end of the closure.
        let enemy_of_my_enemy = {
            let enemy = monster.enemy();
            e.unwrap().name();
        }
        ...
```

rw replied:
Okay, either way is fine by me. I think you should at least add more explanation that *that* particular code is designed intentionally to check a lifetime property. That will help future contributors/maintainers.

mmastrac replied:
I gave it another shot to see if I could explain better - let me know what you think of this! 😄

rw replied:
@mmastrac nice!
